### PR TITLE
Full review of documentation

### PR DESCRIPTION
**Pull request by astrofrog**
This PR exists for the purpose of being able to leave comments on specific parts of the documentation. It will always show the latest master version compared to an empty branch.

### CODE REVIEW

File: python3_user_features.rst [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+Python 3 now allows users to use the ``*variable`` syntax (similar to ``*args``
+in function arguments) to avoid having to write out as many variables than items
+in the iterable::
+
+    >>> a, b, *rest = values
+    >>> a
+    0
+    >>> b
+    1
+    >>> c
```
Comment (dpshelio) [inline python3_user_features.rst]: Should this be `rest` instead of `c`?

File: python3_user_features.rst [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+    >>> value = 4 * 20
+    >>> 'The value is {value}.'.format(value=value)
+    'The value is 80.'
+
+or we can end up in situations where the code is unecessarily complex, since
+``value`` is detached from where it appears in the string.
+
+>>> 'The value is {}.'.format(value)
+'The value is 80.'
```
Comment (dpshelio) [inline python3_user_features.rst]: The indentation of the code blocks is not homogeneous, is that on purpose?

File: python3_user_features.rst (hunk 1 above)
Comment (astrofrog) [inline python3_user_features.rst]: Oops - indeed!

File: python3_user_features.rst (hunk 2 above)
Comment (astrofrog) [inline python3_user_features.rst]: Nope, it needs to be fixed :)

**Comment by dpshelio** [comment]
So far looks really good!! Great job you all!! @astrofrog @stephtdouglas @adrn @ngoldbaum

**Comment by astrofrog** [comment]
@dpshelio - thanks for reviewing! :)

**Comment by bsipocz** [comment]
@astrofrog - I really miss the explicit mention of a bunch of functions (zip, map, etc) that now return iterators rather than lists. Also the methods that now became functions (next, etc).

**Comment by astrofrog** [comment]
@bsipocz - I think we could add more about this to http://python-3-for-scientists.readthedocs.io/en/latest/python3_user_transition_guide.html#other-major-changes-in-python-3 (I don't think it's a feature that's good for the average user, but more a change we should mention in the transition guide)

**Comment by bsipocz** [comment]
@astrofrog - I agree, but it should be mentioned as this is one of the typical things that start to throw errors and bugs when switching, at least I run into them more often than the dict examples listed there. 

Would you add it, or do you want a PR for it?

**Comment by astrofrog** [comment]
@bsipocz - Could you open a PR? Thanks!

File: python3_user_features.rst [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+    >>> import os
+    >>> import glob
+    >>> glob.glob(os.path.join('data', '**', '*.fits'), recursive=True)
+    ['data/image1.fits', 'data/subset1/a.fits', 'data/subset1/b.fits',
+     'data/subset1/c.fits', 'data/subset2/d.fits', 'data/subset2/e.fits']
+
+The ``**`` is used to indicate the point in the path at which to look for
+recursive directories, and the ``recursive=True`` option is needed to
+correctly interpret the ``**``.
```
Comment (jni) [inline python3_user_features.rst]: Whoa! I didn't know about this! =D

File: python3_user_features.rst [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+Since Python 3.5, and Numpy 1.10, it is now possible to use the ``@`` operator
+to do matrix multiplication (vector product)::
+
+    >>> x = np.array([[1, 2], [3, 4]])
+    >>> y = np.array([[3, 2], [2, -1]])
+    >>> x @ y
+    array([[ 7,  0],
+           [17,  2]])
+
```
Comment (jni) [inline python3_user_features.rst]: Perhaps this needs some bigger linear algebra examples to show how much better this makes linear algebra code...

File: python3_user_features.rst [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+functions, to provide information on inputs/outputs. For example, it is possible
+to specify *type* annotations:
+
+    >>> def remove_spaces(x: str) -> str:
+    ...     return x.replace(' ', '')
+
+This syntax means that the input as well as the output should be a string. Now
+it turns out that Python doesn't do anything with these type annotations (there
+are still reasons why developers might want to do this, but this is not
+necessarily critical for the typical user).
```
Comment (jni) [inline python3_user_features.rst]: Note here mypy for static checking of type correctness, and that several IDEs (such as PyCharm) are already using these annotations to help static analysis of your code.

File: python3_user_features.rst (hunk 4 above)
Comment (astrofrog) [inline python3_user_features.rst]: I agree - would you like to contribute a better example?

File: python3_user_features.rst (hunk 5 above)
Comment (astrofrog) [inline python3_user_features.rst]: Cool! Do you have any relevant links we could add here?

Comment (jni) [inline python3_user_features.rst]: https://www.jetbrains.com/help/pycharm/2016.1/type-hinting-in-pycharm.html
http://mypy-lang.org/

File: python3_user_features.rst (hunk 4 above)
Comment (jni) [inline python3_user_features.rst]: A couple of examples from our book:
- Rotating a vector around the z axis:

``` python
import numpy as np

theta = np.deg2rad(45)
R = np.array([[np.cos(theta), -np.sin(theta), 0],
              [np.sin(theta),  np.cos(theta), 0],
              [0,              0,             1]])

print("R @ x-axis:", R @ [1, 0, 0])
print("R @ y-axis:", R @ [0, 1, 0])
print("R @ z-axis:", R @ [0, 0, 1])
```

Our graph layout example is harder to include without lots of context:

``` python
I = sparse.eye(Lap.shape[0], format='csr')
sigma = 0.5

Affn = Dinv2 @ Lap @ Dinv2
v0 = np.ones(Lap.shape[0])

eigvals, vec = sparse.linalg.eigsh(Affn + sigma*I, k=3, which='SM', v0=v0)

sorted_indices = np.argsort(eigvals)
eigvals = eigvals[sorted_indices]
vec = vec[:, sorted_indices]

_ignored, x, y = (Dinv2 @ vec).T
```

File: python3_user_transition_guide.rst [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+To move to Python 3, you can simply replace the imports. To maintain
+compatibility between Python 2 and 3, an easy solution is to use the
+`six <http://pythonhosted.org/six/>`_ package (see discussion below in
+:ref:`maintain-compatibility`).
+
+One other major change is that many built-in methods on container classes (e.g.,
+the dictionary) now return iterators instead of ``list`` s. For example, in
+Python 2::
+
+    >>> airspeed = {'ladened': 2, 'unladened': 11}
```
Comment (jni) [inline python3_user_transition_guide.rst]: I think it's just `laden` and `unladen`. =)

File: python3_user_transition_guide.rst [hunk 7]
```diff
@@ -0,0 +1,243 @@
+Python 3 transition guide
```
Comment (jni) [inline python3_user_transition_guide.rst]: I suggest linking to python-future.org

Also, your own "stop writing Python-4 incompatible code". =)

File: python3_user_features.rst [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+    >>> a, b = 10, 20
+    >>> f'The sum of the values is {a + b}.'
+    'The sum of the values is 30.'
+
+Unicode strings
+---------------
+
+In Python 2, only the basic ASCII character set was available in standard
+strings; to use the much more extensive Unicode set of characters, you had to
+prefix each string with a u:
```
Comment (takluyver) [inline python3_user_features.rst]: This sounds like the only benefit is that you don't need the `u` prefix in Python 3, which is really trivial.

The real benefit is harder to explain: Python 2 lets you casually mix bytes and unicode, which works when you test with `'abc'`, but will probably blow up as soon as your program meets 'Zoë'. Python 3 forces you to think about bytes and text, which is more initial effort, but gives you a better chance of having robust code. This is a tough sell for English speaking scientists who rarely need to think about fancy foreign letters, though.

Comment (astrofrog) [inline python3_user_features.rst]: Maybe we should actually include this as an advanced benefit - the average user might not care/notice? (for typical use cases)

**Comment by astrofrog** [comment]
Closing since I don't think we need to keep this running review open.

**Pull request closed without merging**